src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11KDF.java

```diff
@@ -0,0 +1,363 @@
+/*
+ * Copyright (c) 2024, Red Hat, Inc.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package sun.security.pkcs11;
+
+import javax.crypto.KDFParameters;
+import javax.crypto.KDFSpi;
+import javax.crypto.SecretKey;
+import javax.crypto.spec.HKDFParameterSpec;
+import javax.crypto.spec.SecretKeySpec;
+import java.security.*;
+import java.security.spec.*;
+import java.util.Arrays;
+import java.util.List;
+
+import static sun.security.pkcs11.TemplateManager.*;
+import sun.security.pkcs11.wrapper.*;
+import static sun.security.pkcs11.wrapper.PKCS11Constants.*;
+
+final class P11KDF extends KDFSpi {
+    private final Token token;
+    private final P11SecretKeyFactory.HKDFKeyInfo svcKi;
+    private final long hmacMechanism;
+    private static final SecretKey EMPTY_KEY = new SecretKey() {
+        @Override
+        public String getAlgorithm() {
+            return "Generic";
+        }
+
+        @Override
+        public String getFormat() {
+            return "RAW";
+        }
+
+        @Override
+        public byte[] getEncoded() {
+            return new byte[0];
+        }
+    };
+
+    private static KDFParameters requireNull(KDFParameters kdfParameters,
+            String message) throws InvalidAlgorithmParameterException {
+        if (kdfParameters != null) {
+            throw new InvalidAlgorithmParameterException(message);
+        }
+        return null;
+    }
+
+    P11KDF(Token token, String algorithm, KDFParameters kdfParameters,
+            long hmacMechanism) throws InvalidAlgorithmParameterException {
+        super(requireNull(kdfParameters,
+                algorithm + " does not support parameters"));
+        this.token = token;
+        this.hmacMechanism = hmacMechanism;
+        this.svcKi = P11SecretKeyFactory.getHKDFKeyInfo(algorithm);
+        assert this.svcKi != null : "Only HKDF algorithms supported.";
```

Comment:
The error message is a bit strange, do u mean `"Unsupported HKDF algorithm " + algorithm`?

Reply:
Yes, either way works for me. I'll replace it.